### DISCUSSION TITLE
Default terminal/FM/browser chooser

### DISCUSSION THREAD

**Discussion by AzumaHazuki** [lxqt/lxqt#2331] [Ideas]
Would it be very hard to implement this? Xfce places, in /usr/share/applications/, the files xfce4-{terminal-emulator,file-manager,web_browser}.desktop. These files call "exo-open --launch $UTILITY" when clicked.

Now we obviously don't want to rewrite exo or even parts of it, but would it be possible to have analagous /usr/share/applications/lxqt-$THING.desktop files which, on first run, call a small config dialog to choose preferred defaults and subsequently thereafter run the chosen application?

I would also like to see these three entries at the top of the default application menu, perhaps with a separator and the LXQt Settings submenu directly under them, to make things take fewer clicks to get to.

**Comment by tsujan** [comment]
File Associations → Default Applications

**Reply by AzumaHazuki** [reply]
There is, at least on my system, no entry to choose a terminal, and I'd still like to see Terminal, File Manager, and Web Browser at the top of the main appmenu.

**Reply by tsujan** [reply]
The terminal that's used by pcmanfm-qt is defined in its Preferences dialog. The default terminal doesn't make sense with file association, but it can be set by `qtxdg-mat`; see `qtxdg-mat --help`.

> ... at the top of the main appmenu.

Avoidance of duplication → https://github.com/lxqt/lxqt/discussions/2323#discussioncomment-4095944

Also, the main menu isn't a place for setting default apps.

**Reply by AzumaHazuki** [reply]
Xfce doesn't seem to have a problem doing this, and since we're not locked into one WM (so we can't count on, e.g., a default keybinding to launch these programs...) I'd argue if anything it's even *more* apropos here. People use those three applications very, very commonly and it's a lot quicker and more discoverable.

**Reply by ghost** [reply]
If I'm not mistaken pcmanfm-qt is the only LXQt component with a concept of "Open in terminal". So it makes sense this setting is in pcmanfm-qt and not lxqt-config.

**Reply by tsujan** [reply]
> ... a lot quicker and more discoverable.

Quick Launch has been made for that purpose.

Main menu doesn't and won't have a "favorites" section. There might be a ["Wing Menu"](https://github.com/lxqt/lxqt/discussions/2264#discussioncomment-3167384) if its author makes a PR; it'll have "favorites".

**Reply by stefonarch** [reply]
For the record: another method to tweak the menu, adding favorites or else is editing the menu file.
https://lxqt-project.org/blog/2022/06/18/menu-editing-in-lxqt/

**Reply by tsujan** [reply]
@stefonarch, thanks! I'd forgotten that good page.

**Reply by stefonarch** [reply]
It needs an update, screenshot is from 1.1 :)

**Reply by AzumaHazuki** [reply]
> > ... a lot quicker and more discoverable.
> 
> Quick Launch has been made for that purpose.
> 
> Main menu doesn't and won't have a "favorites" section. There might be a ["Wing Menu"](https://github.com/lxqt/lxqt/discussions/2264#discussioncomment-3167384) if its author makes a PR; it'll have "favorites".

You are misunderstanding me, and perhaps deliberately; while I am definitely waiting with bated breath for WingMenu (and its associated QSS styling directives!) to become part of the panel plugin suite, this is about the standard main menu. Xfce has Terminal, File Manager, and Web Browser as first-class citizens, top-level entries at the top of the default applications menu. 

These are likely the most commonly-used applications by everyone and it makes sense from a usability *and* discoverability perspective to have them there. This is not duplication or bloat; it's three menu entries and a little backend plumbing (which appears to be mostly done already...?) to provide a huge boost in productivity. LXQt has so much potential, and personally I'd like to see it supersede MATE, Cinnamon, and Xfce as *the* light *nix DE. I think it can, too. But it needs to be a little less user-hostile.

**Reply by stefonarch** [reply]
It wouldn't be difficult to write yet another menu file, with` xdg-open ~` for the default filemanager,  but there is no wrapper for the default browser or terminal and `xdg-open` needs an url, but in some way it should be doable.

But bear in mind that there are many ways user do things. I bet usually the browser is already in autostart, I use Super+P for the filemanager and F12 for the terminal, and  I hardly use the menu. Using mostly lxqt-runner for the rest.

**Reply by stefonarch** [reply]
For the default browser this works:
```
[Desktop Entry]
Name=Browser
GenericName=Default browser
Comment=Start default browser
Exec=bash -c "$(basename -s .desktop $(qtxdg-mat def-web-browser))"
Type=Application
Categories=Network;
Icon=web-browser
Terminal=false
```

For some reason here qterminal results not set, although in lxqt-config-session it is.
```
echo $TERM
xterm-256color
```
EDIT: used `qtxdg-mat`)

**Reply by tsujan** [reply]
> You are misunderstanding me, and perhaps deliberately

@AzumaHazuki, I don't have time to "deliberately misunderstand" anyone.

> Xfce has Terminal, File Manager, and Web Browser as first-class citizens...

LXQt doesn't follow LXDE, let alone XFCE or MATE. Every person may have an ideal about how a DE should be, but telling that a DE is "user-hostile" because it doesn't meet his ideal is an overstatement.

**Reply by tsujan** [reply]
> For some reason here qterminal results not set, although in lxqt-config-session it is.

There was a long discussion about it (sorry, I don't remember where). As far as I remember, `TERM` was seen useless.

The default terminal is set by `qtxdg-mat`. See the output of `qtxdg-mat def-terminal`.

**Reply by stefonarch** [reply]
What I'm wondering about are 2 things:
* `$BROWSER` is set in "advanced" and works, but doesn't reflect default browser
* `$TERM` is also set but isn't applied, or overwritten, looked in Lubuntu, same.

**Reply by tsujan** [reply]
Forget about EVs.

`libqtxdg` defines the default browser as the app that handles `text/html`, `x-scheme-handler/http` and `x-scheme-handler/https`. The user can set it in File Associations → Default Applications.

**Reply by stefonarch** [reply]
> Forget about EVs.

Sure, but why do we have them?

**Reply by tsujan** [reply]
I guess they are remainders of old times, before `libqtxdg` could handle the default apps.

I don't think they're needed anywhere now but, maybe, some old-fashioned apps use them. Since they're on the Advanced page, most users can ignore them.

**Reply by stefonarch** [reply]
> The default terminal is set by `qtxdg-mat`. See the output of `qtxdg-mat def-terminal`.

Btw I always forget about this, and it's not even in the wiki afaik, needs a paragraph in https://github.com/lxqt/lxqt/wiki/ConfigMiscellaneous

**Reply by tsujan** [reply]
Yes. `qtxdg-tools` is a relatively new component. Sadly, [Luis' patch](https://gitlab.freedesktop.org/xdg/xdg-utils/-/merge_requests/48) hasn't been merged into `xdg-utils` yet.

**Reply by stefonarch** [reply]
Added https://github.com/lxqt/lxqt/wiki/ConfigMiscellaneous#default-applications - could you check it please @tsujan ?

**Reply by tsujan** [reply]
Thanks for adding a section for it! I checked and edited it.

**Comment by stefonarch** [comment]
I played a little bit, that what is proposed here can be achieved with a text editor, copying `lxqt-applications.menu` to `/etc/xdg/menu/defapps.menu`:

At the bottom after Screensaver change to:
```

	</Menu> <!-- End Screensaver -->

		<!-- Default apps -->
	<Menu>
		<Name>Defapps</Name>
		<Include>
		<Filename>browser.desktop</Filename>
		<Filename>email.desktop</Filename>
		<Filename>filemanager.desktop</Filename>
		<Filename>terminal.desktop</Filename>
		</Include>
	</Menu> <!-- End default apps -->


	<Layout>
		<Merge type="files"/>
		<Merge type="menus"/>
		<Separator/>
		<Menuname>DesktopSettings</Menuname>
		<Menuname show_empty="false" inline="true">LXQt-About</Menuname>
		<Separator/>
		<Menuname show_empty="false" inline="true">Defapps</Menuname>
		<Menuname show_empty="false">X-Leave</Menuname>
		<Menuname show_empty="false" inline="true">Screensaver</Menuname>
	</Layout>

</Menu> <!-- End Applications -->

```

Only  items for which the requested desktop exists (e.g. in `~/.local/share/applications`) will be shown.
Example for `browser.desktop`:
```

[Desktop Entry]
Name=Browser
GenericName=Default browser
Comment=Start default browser
Exec=bash -c "$(basename -s .desktop $(qtxdg-mat def-web-browser))"
Type=Application
Categories=Network;
Icon=web-browser
Terminal=false
```

Possibile is showing those:
 ```
  def-web-browser   Get/Set the default web browser
  def-email-client  Get/Set the default email client
  def-file-manager  Get/Set the default file manager
  def-terminal      Get/Set the default terminal
```
![screen_area_mer_17:18:57_](https://user-images.githubusercontent.com/10681413/202239528-e21c2930-5836-4dec-8eab-82c5c47795be.png)

**Reply by stefonarch** [reply]
I'm not sure if this should be implemented, as the wing menu covers much better those things.

**Comment by AzumaHazuki** [comment]
I have a slightly modified version of the above, which relies on lxqt-{terminal,filemanager,browser}.desktop entries. It mostly works, but 1) why can we not simply place <Filename></Filename> entries in the Layout block and 2) _why_ do they insist on alphabetizing by the .desktop entries' Name= strings instead of...you know...displaying them in the order specified?

![defaultapps_000](https://user-images.githubusercontent.com/67122280/202351200-cd9567bf-6ab5-4469-a302-aff3e091e700.jpg)

**Reply by stefonarch** [reply]
1+2) because this is how it's designed.

To achieve your preferred ordering you could try using a menu for each of those 3 .desktop files, like

```
<Menu>
		<Name>Browser</Name>
		<Include>
		<Filename>browser.desktop</Filename>
		</Include>
</Menu>
```

**Reply by ghost** [reply]
I wonder if the alphabetized order is an XDG standard and/or implemented in other DEs? It works well because separate .menu files are kept consistent, but could be changed if backwards compatibility with old custom .menu's is given up.

I don't know about the \<Menu\> and \<Layout\> distinction.

**Reply by stefonarch** [reply]
https://specifications.freedesktop.org/menu-spec/menu-spec-latest.html#introduction

**Reply by AzumaHazuki** [reply]
> 1+2) because this is how it's designed.
> 
> To achieve your preferred ordering you could try using a menu for each of those 3 .desktop files, like
> 
> ```
> <Menu>
> 		<Name>Browser</Name>
> 		<Include>
> 		<Filename>browser.desktop</Filename>
> 		</Include>
> </Menu>
> ```

That's...okay, I see why it has to be that way. It's ugly, but it's consistent. So, here's what I've come up with so far, and the edits to the standard menu to make it happen:

![LXQt-Custom](https://user-images.githubusercontent.com/67122280/202594354-80bd23c9-290e-4dbd-b3e2-e9326dafc937.jpg)

**Comment by stefonarch** [comment]
> I would also like to see these three entries at the top of the default application menu, perhaps with a separator and the LXQt Settings submenu directly under them, to make things take fewer clicks to get to.

My 2 cents finally:

* Many people do no use a terminal or an email application
* it's not possbile to make it configurable easily
* translations would be needed (doable)
* quicklaunch-plugin is one click less still
* wing menu will cover all of it

**Reply by AzumaHazuki** [reply]
Do we plan to default to Wing Menu, then? I wouldn't mind that at all, but even then, we're still at square one: no defaults set for these apps.

**Reply by stefonarch** [reply]
We hope to get a PR soon for wing menu - if it will be default or not we'll see, anyway distributions ship their `panel.conf` with the plugins  they want it.

Defaults are set in `lxqt-config-file-associations `.
 I don't know if shipping general `.desktop` files for those 4 applications as above is a good idea. People open firefox or chromium or falkon  not "browser".

**Reply by ghost** [reply]
>Do we plan to default to Wing Menu, then? 
>We hope to get a PR soon for wing menu - if it will be default or not we'll see

It is good to improve old plugins too, the code complexity for editing the .menu files is very minor - if we want the .desktop files. As far as total code/configuration complexity goes, it isn't that bad.

**Reply by stefonarch** [reply]
Power users will be able to do anything with menu files, but average users don't even change them, nor will they find and remove `.desktop` files to hide email or terminal entry.

**Reply by ghost** [reply]
Is there a way to embed a command in a XDG menu?

**Reply by stefonarch** [reply]
No, afaik, only using .desktop files.

**Reply by AzumaHazuki** [reply]
> > Do we plan to default to Wing Menu, then?
> > We hope to get a PR soon for wing menu - if it will be default or not we'll see
> 
> It is good to improve old plugins too, the code complexity for editing the .menu files is very minor - if we want the .desktop files. As far as total code/configuration complexity goes, it isn't that bad.

Maybe we're overthinking this a bit: a dialog like the WM chooser, with an associated list of WMs like that one has, would be perfect for this, and it could create the lxqt-default-$THING.desktop files as *symlinks* to the chosen applications' .desktop files. Since the menu spec has an option not to show an empty entry (i.e., no .desktop exists for it), there's no danger of the menu showing, for example, a Thunderbird entry that doesn't do anything if the user doesn't select an email client.

Here are the terminal, filemanager, and browser equivalents of /usr/share/lxqt/windowmanagers.conf I've come up with:

```
[KnownBrowsers]

chromium/Name=chromium
chromium/Comment=Open-source version of Google's Chrome browser

epiphany/Name=epiphany
epiphany/Comment=GNOME Project's GTK 3/Webkit-based browser

falkon/Name=falkon
falkon/Comment=Lightweight, feature-rich Qt-based browser

firefox/Name=firefox
firefox/Comment=Mozilla's Firefox web browser

google-chrome/Name=google-chrome
google-chrome/Comment=The web browser from Google

microsoft-edge/Name=microsoft-edge
microsoft-edge/Comment=Chromium-based web browser from Microsoft

opera/Name=opera
opera/Comment=A fast, secure web browser

otter/Name=otter
otter/Comment=An attempt to recreate classic Opera (12.x) experience in Qt5

seamonkey/Name=seamonkey
seamonkey/Comment=Mozilla's Seamonkey browser and application suite
```
```
[KnownTerminals]

qterminal/Name=qterminal
qterminal/Comment=Light, Qt-based terminal; LXQts default. Choose this if uncertain.

alacritty/Name=alacritty
alacritty/Comment=Modern, GPU-accelerated terminal emulator

aterm/Name=aterm
aterm/Comment=RXVT-compatible terminal emulator with transparency support

gnome-terminal/Name=gnome-terminal
gnome-terminal/Comment=GNOME Project's terminal emulator

kitty/Name=kitty
kitty/Comment=Fast, feature-rich, GPU-accelerated terminal emulator

konsole/Name=konsole
konsole/Comment=KDE's terminal emulator

lilyterm/Name=lilyterm
lilyterm/Comment=Standalone terminal based on libvte that aims for speed and light weight

mate-terminal/Name=mate-terminal
mate-terminal/Comment=MATE DE's terminal emulator

mlterm/Name=mlterm
mlterm/Comment=A multi-lingual terminal emulator

roxterm/Name=roxterm
roxterm/Comment=Terminal emulator for the ROX Desktop Environment

rxvt-unicode/Name=rxvt-unicode
rxvt-unicode/Comment=RXVT clone with Perl extensions, Xft, and Unicode support

sakura/Name=sakura
sakura/Comment=Lightweight GTK 3/VTE terminal emulator

st/Name=st
st/Comment=Ultra-minimalist terminal emulator from the Suckless Project

terminator/Name=terminator
terminator/Comment=Multiple GNOME Terminals in one window!

terminology/Name=terminology
terminology/Comment=Enlightenment's terminal emulator, based on EFL toolkit

xfce4-terminal/Name=xfce4-terminal
xfce4-terminal/Comment=Xfce Desktop's terminal emulator, built on GTK 3 and VTE

xterm/Name=xterm
xterm/Comment=The default X11/Xorg terminal emulator
```
```
[KnownFilemanagers]

pcmanfm-qt/Name=pcmanfm-qt
pcmanfm-qt/Comment=LXQt's default file manager, based on Qt. Choose this if unsure.

caja/Name=caja
caja/Comment=MATE DE's fork of GNOME Nautilus

dolphin/Name=dolphin
dolphin/Comment=KDE's file manager

krusader/Name=krusader
krusader/Comment="Orthodox" (twin-panel) file manager written in Qt

nautilus/Name=nautilus
nautilus/Comment=GNOME Project's file manager

nemo/Name=nemo
nemo/Comment=Cinnamon DE's fork of GNOME Nautilus

pcmanfm/Name=pcmanfm
pcmanfm/Comment=Older, GTK-based file manager from LXDE project

spacefm/Name=spacefm
spacefm/Comment=Multi-panel, tabbed file manager written in GTK

thunar/Name=thunar
thunar/Comment=Xfce's default file manager
```

**Reply by stefonarch** [reply]
If (if) this should be implemented it would need only one thing (beside the updated `.menu` file): in the mainmenu settings a list of the default applications (4) with a checkbox for each "Display in menu"  which will create the `.desktop` files. And then again it depends of the position of the panel. 

And you need a developer who likes the idea and makes a PR.

**Reply by AzumaHazuki** [reply]
So you're envisioning this being part of the Main Menu widget configuration sub-dialog then? That's clever :) I'd still like to see a few pages added to the first run screen like for the WM selection but using the above, for those distros that don't ship a default config.

I'd do this all myself but, and sorry to play to the stereotypes about women and computers here, I cannot code, like *at all.* @_@ So it seems like all I can really contribute to LXQt is themes and mockups and some ideas.

**Reply by stefonarch** [reply]
If  for each feature request or idea would be created a coder we would be a BIG DE ;)
But you'r not alone, my coding skills are very limited too.

**Comment by isf63** [comment]
Is there a way to have a separator in a `<Layout>` only if there exist >0 desktop files in a `<Merge type="files"\>`? This would be nice for the Preferences menu in `lxqt-applications.menu` where LXQt settings run into user settings.

https://github.com/lxqt/lxqt-menu-data/blob/229b9c443c123524a37bdd84c1f4ff0e6a082858/menu/lxqt-applications.menu#L163
